src/ring_theory/dedekind_domain/pid.lean

```diff
@@ -0,0 +1,187 @@
+/-
+Copyright (c) 2023 Anne Baanen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Anne Baanen
+-/
+
+import ring_theory.dedekind_domain.dvr
+import ring_theory.dedekind_domain.ideal
+
+/-!
+# Proving a Dedekind domain is a PID
+
+This file contains some results that we can use to show all ideals in a Dedekind domain are
+principal.
+
+## Main results
+
+ * `is_principal_ideal_ring.of_finite_primes`: if a Dedekind domain has finitely many prime ideals,
+   it is a principal ideal domain.
+-/
+
+variables {R : Type*} [comm_ring R] [is_domain R] [is_dedekind_domain R]
+
+open ideal
+open unique_factorization_monoid
+open_locale big_operators
+open_locale non_zero_divisors
+
+open unique_factorization_monoid
+
+/-- Let `P` be a prime ideal, `x ∈ P \ P²` and `x ∉ Q` for all prime ideals `Q ≠ P`.
+Then `P` is generated by `x`. -/
+lemma ideal.eq_span_singleton_of_mem_of_not_mem_sq_of_not_mem_prime_ne
+  {P : ideal R} (hP : P.is_prime)
+  {x : R} (x_mem : x ∈ P) (hxP2 : x ∉ P^2)
+  (hxQ : ∀ (Q : ideal R), is_prime Q → Q ≠ P → x ∉ Q) :
+  P = ideal.span {x} :=
+begin
+  letI := classical.dec_eq (ideal R),
+  have hx0 : x ≠ 0,
+  { rintro rfl,
+    exact hxP2 (zero_mem _) },
+  by_cases hP0 : P = ⊥,
+  { unfreezingI { subst hP0 },
+    simpa using hxP2 },
+  have hspan0 : span ({x} : set R) ≠ ⊥,
+  { refine mt ideal.span_eq_bot.mp _,
+    simpa only [not_forall, set.mem_singleton_iff, exists_prop, exists_eq_left] using hx0 },
+  have span_le := (ideal.span_singleton_le_iff_mem _).mpr x_mem,
+  refine associated_iff_eq.mp
+    ((associated_iff_normalized_factors_eq_normalized_factors hP0 hspan0).mpr
+      (le_antisymm ((dvd_iff_normalized_factors_le_normalized_factors hP0 hspan0).mp _) _)),
+  { rwa [ideal.dvd_iff_le, ideal.span_singleton_le_iff_mem] },
+  simp only [normalized_factors_irreducible ((ideal.prime_of_is_prime hP0 hP).irreducible),
+      normalize_eq, multiset.le_iff_count, multiset.count_singleton],
+  intros Q,
+  split_ifs with hQ,
+  { unfreezingI { subst hQ },
+    refine (ideal.count_normalized_factors_eq _ _).le;
+      simp only [ideal.span_singleton_le_iff_mem, pow_one];
+      assumption },
+  by_cases hQp : is_prime Q,
+  { resetI,
+    refine (ideal.count_normalized_factors_eq _ _).le;
+      simp only [ideal.span_singleton_le_iff_mem, pow_one, pow_zero, one_eq_top, submodule.mem_top],
+    exact hxQ _ hQp hQ },
+  { exact (multiset.count_eq_zero.mpr (λ hQi, hQp (is_prime_of_prime (irreducible_iff_prime.mp
+      (irreducible_of_normalized_factor _ hQi))))).le }
+end
+
+/-- A Dedekind domain is a PID if its set of primes is finite. -/
+theorem is_principal_ideal_ring.of_finite_primes
+  (h : set.finite {I : ideal R | is_prime I}) :
+  is_principal_ideal_ring R :=
```

Comment:
I formalized a direct proof [here](https://gist.github.com/alreadydone/a42fdaab98f5005dc4c2ba523fea1d5e) that proves something slightly more general, but it's surprisingly painful. Certainly there are lemmas to be extracted but those wouldn't make it much less painful ...